### PR TITLE
Ensure that Progress is not nil when a MsgSnapStatus comes in

### PR DESCRIPTION
**Pull request by BramGruneir**
This was causing some issues in cockroach cockroachdb/cockroach#2950

### CODE REVIEW

**Comment by BramGruneir** [comment]
attn: @bdarnell

File: raft/raft.go [hunk 1]
```diff
@@ -632,7 +632,7 @@ func stepLeader(r *raft, m pb.Message) {
 			r.id, r.raftLog.lastTerm(), r.raftLog.lastIndex(), r.Vote, m.From, m.LogTerm, m.Index, r.Term)
 		r.send(pb.Message{To: m.From, Type: pb.MsgVoteResp, Reject: true})
 	case pb.MsgSnapStatus:
-		if pr.State != ProgressStateSnapshot {
+		if pr == nil || pr.State != ProgressStateSnapshot {
```
Comment (xiang90) [inline raft/raft.go]: probably you want to add this check at line 556 for the msg that access fields of pr. then you can also remove line 569.

Comment (BramGruneir) [inline raft/raft.go]: I'd love to just move the check at 569 up, but that's looking for r.id and not m.From

Also, only 5 out of the 8 messages types, look at pr.
- MsgProp
- MsgAppResp
- MsgHeartbeatResp
- MsgSnapStatus
- MsgUnreachable

While these don't:
- MsgVote
- MsgCheckQuorum
- MsgBeat

I'll add the check for each of the ones that rely on it.

Comment (xiang90) [inline raft/raft.go]: Yea. I think that is up to you. I just want to find a more consistent way to check the nil value.

Comment (bdarnell) [inline raft/raft.go]: `MsgBeat` and `MsgCheckQuorum` are internally-generated messages with `m.From == r.id`. If `r.id` is not in `r.prs`, I think it should be fine to drop those messages too. Then `MsgVote` is the only message that a leader may need to handle from a node not in `r.prs`. 

`MsgVote` handling is split between the top-level `Step` method and the per-state `step{Leader,Candidate,Follower}`. If we could move the sending of rejections from `stepLeader` to `Step`, we could do the `r.prs` check at the start of `stepLeader` no matter what the message type is.

Comment (BramGruneir) [inline raft/raft.go]: I tried pulling out the voting check, but MsgBeat still needs to be able to run without progress.  So I've instead split the switch statement in two.  With the first set of message types that do not require that the progress exist, the second set requiring that it does exists, and the check for it in the middle.

**Comment by BramGruneir** [comment]
I think this is good to go.  PTAL

**Comment by bdarnell** [comment]
LGTM

File: raft/raft.go [hunk 2]
```diff
@@ -582,6 +582,20 @@ func stepLeader(r *raft, m pb.Message) {
...
+		r.send(pb.Message{To: m.From, Type: pb.MsgVoteResp, Reject: true})
+	}
+
+	pr, prOk := r.prs[m.From]
+	if !prOk {
+		r.logger.Debugf("no progress available for %x", m.From)
+		return
+	}
+
+	// These messages types do require progress for m.From (pr).
```
Comment (xiang90) [inline raft/raft.go]: let move this line before the pr checking at line 591?

**Comment by xiang90** [comment]
@BramGruneir 

The approach looks good to me. Let's fix the comment issue and change the commit message to 
```
raft: Ensure that Progress is not nil when a MsgSnapStatus comes in
```

File: raft/raft.go (hunk 2 above)
Comment (BramGruneir) [inline raft/raft.go]: sure, done.

**Comment by BramGruneir** [comment]
commit message changed, comment moved and slightly reworded.

File: raft/raft.go [hunk 3]
```diff
@@ -582,6 +582,19 @@ func stepLeader(r *raft, m pb.Message) {
 		}
 		r.appendEntry(m.Entries...)
 		r.bcastAppend()
+	case pb.MsgVote:
+		r.logger.Infof("%x [logterm: %d, index: %d, vote: %x] rejected vote from %x [logterm: %d, index: %d] at term %d",
+			r.id, r.raftLog.lastTerm(), r.raftLog.lastIndex(), r.Vote, m.From, m.LogTerm, m.Index, r.Term)
+		r.send(pb.Message{To: m.From, Type: pb.MsgVoteResp, Reject: true})
+	}
+
```
Comment (xiang90) [inline raft/raft.go]: I think we should probably return here if we hit any of the cases in the previous switch.

Comment (xiang90) [inline raft/raft.go]: or we might see the unnecessary debugging line below at line594.

Comment (BramGruneir) [inline raft/raft.go]: Oh, good point.  I'll add a return to each of the above cases.

Comment (BramGruneir) [inline raft/raft.go]: done

**Comment by xiang90** [comment]
LGTM

**Comment by BramGruneir** [comment]
Thanks @xiang90 !

**Pull request merged**